From db62e058dc6542bc3c6fc8d55492a432d12e0308 Mon Sep 17 00:00:00 2001
From: "Huang.Alang" <huang.alang@inventec.corp-partner.google.com>
Date: Wed, 3 May 2023 00:49:37 +0000
Subject: [PATCH] superion machine init

<details>
- ast snoop port80 to sgpioA configuration
- enable SAFS (slave attached flash sharing mode)

%% original patch: 0002-superion-machine-init.patch
---
 arch/arm/mach-aspeed/ast2600/board_common.c | 90 ++++++++++++++++++++-
 1 file changed, 89 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-aspeed/ast2600/board_common.c b/arch/arm/mach-aspeed/ast2600/board_common.c
index 70d2a97b5d..98b24c93fd 100644
--- a/arch/arm/mach-aspeed/ast2600/board_common.c
+++ b/arch/arm/mach-aspeed/ast2600/board_common.c
@@ -13,6 +13,39 @@
 #include <dm/uclass.h>
 
 DECLARE_GLOBAL_DATA_PTR;
+/* LPC Snoop related Registers*/
+#define SCU_BASE                        0x1e620000
+#define GPIO_BASE                       0x1e780000
+#define SGPIOM                          SCU_BASE + 0x414
+#define AST_LPC_BASE                    0x1e789000
+#define SNPWADR                         0x090   /* LPC Snoop Address Register */
+#define SNPWDR                          0x094   /* LPC Snoop Data Register */
+#define LPC_SNOOP_ADDR                  0x80
+#define HICR6                           0x084   /* Host Interface Control Register 6 */
+#define HICR5                           0x080   /* Host Interface Control Register 5 */
+/* HRCR6 Bits */
+#define HICR6_STR_SNP0W                 (1 << 0) /* Interrupt Status Snoop address 0 */
+#define HICR6_STR_SNP1W                 (1 << 1) /* Interrupt Status Snoop address 1 */
+#define HICR5_UNKVAL_MASK               0x1FFF0000 /* Bits with unknown values on reset */
+#define HICR5_EN_SIOGIO                 (1 << 31)
+#define HICR5_EN_SNP0W                  (1 << 0)
+#define GPIO554                         0x554
+#define HICRB                           0x100
+
+/* HICRB Bits */
+#define HICRB_EN80HSGIO                 (1 << 13) /* Enable 80hSGIO */
+#define SGPIO_CLK_DIV(N)                ((N) << 16)
+#define SGPIO_BYTES(N)                  ((N) << 6)
+#define SGPIO_ENABLE                    1
+
+/* Hardware Strap2 Registers*/
+#define SCU510_HW_STRAP2	(SCU_BASE + 0x510)
+#define BIT_SAFS_MODE       (BIT(7))
+#define GPIO_BASE           (0x1e780000)
+#define GPIOQ_DATA          (GPIO_BASE + 0x080)
+#define GPIOQ_DIR           (GPIO_BASE + 0x084)
+#define BIT_GPIOQ1          (BIT(1))
+#define BIT_GPIOQ3          (BIT(3))
 
 /*
  * RMII daughtercard workaround
@@ -77,7 +110,9 @@ __weak int board_init(void)
 #ifdef ASPEED_RMII_DAUGHTER_CARD
 	reset_eth_phy();
 #endif
-	/*
+    sgpio_init_port80();
+    safs_enable();
+    /*
 	 * Loop over all MISC uclass drivers to call the comphy code
 	 * and init all CP110 devices enabled in the DT
 	 */
@@ -227,3 +262,56 @@ void aspeed_mmc_init(void)
 	return;
 
 }
+
+int sgpio_init_port80(void) {
+
+    u32 reg;
+
+    //set SGPIO mutifunctin pin
+    reg = readl(SGPIOM);
+    reg |= (0x0F000000);
+    writel(reg,SGPIOM);
+
+    reg = readl(AST_LPC_BASE + SNPWADR) & 0xffff0000;
+    writel(reg | LPC_SNOOP_ADDR, AST_LPC_BASE + SNPWADR);
+
+    /* clear interrupt status */
+    reg = readl(AST_LPC_BASE + HICR6);
+    reg |= HICR6_STR_SNP0W | HICR6_STR_SNP1W;
+    writel(reg, AST_LPC_BASE + HICR6);
+
+    /* enable lpc snoop #0 and SIOGIO */
+    reg = readl(AST_LPC_BASE + HICR5) & ~(HICR5_UNKVAL_MASK);
+    reg |= HICR5_EN_SIOGIO | HICR5_EN_SNP0W;
+    writel(reg, AST_LPC_BASE + HICR5);
+
+
+    /* enable port80h snoop on SGPIO */
+    reg = readl(AST_LPC_BASE + HICRB) | HICRB_EN80HSGIO;
+    writel(reg, AST_LPC_BASE + HICRB);
+
+    /* set the gpio clock to pclk/(2*(5+1)) or ~2 MHz */
+    reg = SGPIO_CLK_DIV(99) | SGPIO_BYTES(1) | SGPIO_ENABLE;
+    writel(reg, GPIO_BASE + GPIO554);
+
+    return 0;
+}
+
+void safs_enable(void)
+{
+    u32 reg = 0;
+    /*SCU510 setting*/    
+    reg = readl(SCU510_HW_STRAP2);
+    reg |= BIT_SAFS_MODE;
+    writel(reg, SCU510_HW_STRAP2);
+    /*gpio setting*/
+    //set GPIO Q1: SPI_BIOS_MUXA_SEL Q3: SPI_BIOS_MUXB_SEL 
+    //output high
+    reg = readl(GPIOQ_DATA);
+    reg = reg | BIT_GPIOQ1 | BIT_GPIOQ3;
+    writel(reg, GPIOQ_DATA);
+
+    reg = readl(GPIOQ_DIR);
+    reg = reg | BIT_GPIOQ1 | BIT_GPIOQ3;
+    writel(reg, GPIOQ_DIR);
+}    
-- 
2.17.1

